docs/registries.1
.TH "REGISTRIES" "1" " " "Brent Baude" "May 2017" 
.nh
.ad l


.SH NAME
.PP
registries \- parses a YAML configuration file for registries


.SH SYNOPSIS
.PP
\fBregistries\fP
[\fB\-h\fP|\fB\-\-help\fP]
[\fB\-i\fP][\fB\-\-input\fP]
[\fB\-j\fP][\fB\-\-json\fP]
[\fB\-o\fP][\fB\-\-output\fP]
[\fB\-V\fP][\fB\-\-variable\fP]


.SH DESCRIPTION
.PP
\fBregistries\fP parses a YAML configuration file about registries and has the
ability to output in string or JSON formats.  By default, \fB\fCregistries\fR will
put its output to stdout in string format.  The output is primarily for other
programs.


.SH OPTIONS:
.PP
\fB\-h\fP \fB\-\-help\fP
Print usage statement

.PP
\fB\-i\fP \fB\-\-input\fP
Allows you to specify a different path and name for a YAML input file.  It
overrides the default configuration file of \fB\fC/etc/containers/registries.conf\fR\&.

.PP
\fB\-j\fP \fB\-\-json\fP
Specifies the output be in JSON format.  The default is to output as a string.

.PP
\fB\-o\fP \fB\-\-output\fP
Redirects the output from stdout to a file. This option is useful when your
program requires its input come from a file (i.e. Systemd Environment files).

.PP
\fB\-V\fP \fB\-\-variable\fP
This option allows you to specify a variable name in the final output.  The contents
of the variable will also be quoted.  It can only be used when writing to
an output file.  For example, a \fB\fC\-\-variable\fR of
\fB\fCREGISTRIES\fR will result in the output format to be like:

.PP
.RS

.nf
     REGISTRIES="OUTPUT"

.fi
.RE

.PP
This option is useful for putting the contents into a SystemD Environment file
for use by a SystemD service.


.SH EXAMPLES
.PP
To parse the default file location and obtain a result to stdout in string
format:

.PP
.RS

.nf
# registries

.fi
.RE

.PP
To parse the default file location and obtain a result to stdout as JSON:

.PP
.RS

.nf
# registries \-j

.fi
.RE

.PP
To parse the default file location and write the results to a file and insert
a variable that can be exported.

.PP
.RS

.nf
# registries \-o /tmp/foo \-V registries

.fi
.RE

.PP
You can also use the \-i command line switch to read in a file other than the
default input file:

.PP
.RS

.nf
# registries \-i /tmp/new\_file

.fi
.RE


.SH HISTORY
.PP
May 2017, Originally compiled by Brent Baude (bbaude at redhat dot com)


registries/registries.py
# System-wide registries parsing tool

import sys
from abc import abstractmethod, ABCMeta
import yaml
import pytoml
import argparse
import json


map_output = {
    "registries.search": "--add-registry",
    "registries.insecure":  "--insecure-registry",
    "registries.block": "--block-registry"
}


def write_file(filename, data):
    f = open(filename,"w")
    f.write(data)
    f.close()


def to_dict(secure, insecure, block):
    return {
        "registries.search": {"registries": secure["registries"]},
        "registries.insecure": {"registries": insecure["registries"]},
        "registries.block": {"registries": block["registries"]}
}


def do_output(output, args):
    if args.output:
        if args.variable:
            output = "{}=\"".format(args.variable) + output + "\""
        write_file(args.output, output)
    else:
        sys.stdout.write("{}\n".format(output))


def to_string(config):
    out_str = ""
    for i in map_output:
        for registry in config[i]['registries']:
            out_str += "{} {} ".format(map_output[i], registry)
    return out_str


class Conf(object): #pylint: disable=metaclass-assignment
    # Mark the class as abstract
    __metaclass__ = ABCMeta

    @abstractmethod
    def load(registries_conf_file):
        pass

    @abstractmethod
    def get_registries(self, conf_file):
        pass


class LoadError(Exception):
    pass


class loadYAML(Conf):
    @staticmethod
    def load(registries_conf_file):
        with open(registries_conf_file, 'r') as stream:
            return yaml.load(stream)

    @classmethod
    def get_registries(cls, conf_file):
        config = cls.load(conf_file)
        _registries = [] if "registries" not in config else config["registries"]
        _insecure= [] if "insecure_registries" not in config else config['insecure_registries']
        _block = [] if "block_registries" not in config else config['block_registries']
        return to_dict({"registries": _registries}, {"registries": _insecure}, {"registries": _block})


class loadTOML(Conf):
    @staticmethod
    def load(registries_conf_file):
        with open(registries_conf_file, 'rb') as stream:
            return pytoml.load(stream)

    @classmethod
    def get_registries(cls, conf_file):
        config = cls.load(conf_file)
        reg = config['registries']
        _registries = [] if 'search' not in reg else reg['search']
        _insecure = [] if 'insecure' not in reg else reg['insecure']
        _block = [] if "block" not in reg else reg['block']
        return to_dict(_registries, _insecure, _block)


def load_config(registries_conf_file):
    try:
        registries = loadTOML.get_registries(registries_conf_file)
    except (pytoml.TomlError, KeyError):
        try:
            registries = loadYAML.get_registries(registries_conf_file)
        except yaml.YAMLError:
            sys.stderr.write("Unable to load and parse {}.".format(registries_conf_file))
            sys.exit(1)

    return registries


def registries():
    parser = argparse.ArgumentParser()
    parser.add_argument("-i", "--input", help="Specify an input file", default="/etc/containers/registries.conf")
    parser.add_argument("-j", "--json", help="Output in JSON format", action="store_true")
    parser.add_argument("-o", "--output", help="Specify an output file")
    parser.add_argument("-V", "--variable", help="Specify a vcariable assignment")
    args = parser.parse_args()

    if args.variable and not args.output:
        sys.stderr.write("You must use --output with -V")
        sys.exit(1)

    registries_config = load_config(args.input)
    output = json.dumps(registries_config) if args.json else to_string(registries_config)


    do_output(output, args)

if __name__ == '__main__':
    registries()